### ISSUE TITLE
Jest 27 errors with `ReferenceError: setImmediate is not defined`

### ISSUE THREAD

**Issue by ericwooley** [prisma/prisma#8558]
## Problem

While upgrading to 27.0.3 of jest, I have failing integration tests because 

```
 ● Test suite failed to run

    ReferenceError: setImmediate is not defined

      at Parser.destroy (../../../node_modules/@prisma/client/runtime/index.js:25498:7)
      at detachSocket (../../../node_modules/@prisma/client/runtime/index.js:25536:21)
      at Socket.onSocketClose (../../../node_modules/@prisma/client/runtime/index.js:25545:5)
```

## Suggested solution

 Because [setImmediate](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate) is a non-standard API in the first place, and only officially supported by ie10, prisma should remove references to it.
 
 
Sorry if this is the wrong report time, I was split between a feature request and a bug.

**Comment by janpio** [comment]
> While upgrading to 27.0.3

Version `27.0.3`of what exactly?

What node version are you using?

**Comment by ericwooley** [comment]
Hi @janpio 

Sorry, 27.0.3 of jest

14.17.4

I didn't realize it was built into node, I'll have to investigate why jest 27 is breaking that

**Comment by ericwooley** [comment]
For anyone else getting here from google or whatever, there are lots of changes regarding setImmediate in the jest changelog for 27 https://github.com/facebook/jest/blob/master/CHANGELOG.md

I'm still investigating, but I suspect https://github.com/facebook/jest/pull/11222 might have to do with it.

obviously make sure your test env is set to node. (ours is, and we still get the problem :man_shrugging: ). I'll update this issue when I figure it out.

**Comment by erawk** [comment]
@ericwooley @janpio I've seen this error as well. Downgrading from Jest `27.x` to `26.6.3` solved the problem.

The error that was being thrown was:

```
    ReferenceError: setImmediate is not defined

      at Parser.destroy (node_modules/@prisma/client/runtime/index.js:27664:7)
```

And in the generated runtime client, the code that calls `setImmediate` looks like:

```javascript
    destroy() {
      clearTimeout(this.timeout);
      this.timeout = null;
      this.unconsume();
      setImmediate((self) => self.close(), this);
    }
```

Downgrading Jest hid this error, likely due to Jest overriding `setImmediate` to muck around with timers internally.

Using `jest.useFakeTimers()` in `jest.setup.ts` hid the issue, but if we call `prisma.$disconnect()` to clean up, we see the issue again. Running `jest.runAllTimers()` in an `afterAll` callback after we call `prisma.$disconnect()` did not help either.

This is likely a Jest issue, but not sure if there's any advice on how to avoid triggering this.

**Comment by gendronb** [comment]
@erawk I'm using Jest `26.6.3` and I still get the error...

**Comment by TomSssM** [comment]
Same problem here, using jest + ts-jest, had to downgrade jest and the related deps to major version 26:
```json
{
  "@types/jest": "26.0.14",
  "jest": "26.6.3",
  "ts-jest": "26.5.6"
}
```

**Comment by ericwooley** [comment]
Just an update as to my investigation on this. I never figured it out, I think updating to an even newer jest version than 27.0.3  may have fixed it, but I don't have access to the codebase where this was a problem anymore. So I won't be able to investigate this.

**Comment by davidmoshal** [comment]
Some problem, using Playwright

**Comment by TomSssM** [comment]
By the way, make sure your env is set to `node` instead of `jsdom` because `setImmediate` doesn't exist in `jsdom` as it is Node-only feature

**Comment by davidmoshal** [comment]
@TomSssM: that's a very strange response. 
I need jsdom because, as stated above, I'm using Playwright (a browser automation framework).

Jest 26 worked fine (as you noticed), but Jest 27 fails, even if you set the test environment to jsdom, 
and add the relevant dependencies.

It's curious to me that Jest 26 works (with the default jsdom environment),
yet Jest 27 fails when you explicitly set the test environment to jsdom.

I wonder what has changed in the Jest 27 jsdom implementation?

**Comment by TomSssM** [comment]
@davidmoshal my theory was that env `jsdom` supposes you to run client-side only code inside of it and client side code should not use `setImmediate` as it is a Node.js feature. When I changed my env to `node` (after accidentally setting it to `jsdom`) everything started working, when env is `jsdom` you don't have `setImmediate` in the global object. I guess the reasoning behind it is what I wrote above (client side code should not rely on Node.js specific features). But hey, I might be wrong, I was simply debugging why my jest tests were failing (alas the bug was caused by my lack of attention).

**Comment by davidmoshal** [comment]
@TomSssM , you're missing the point.
I'm testing browsers, ie: I'm testing client side code, therefore I need Jsdom.
If I set it to 'node' then it complains about numerous things, eg: no window object, etc.
Take a look at Playwright: https://playwright.dev/
Maybe that will help you understand?

**Comment by TomSssM** [comment]
@davidmoshal I already understood your point and I know that switching from `jsdom` to `node` is not an option for you, in my initial reply I simply wrote what helped me out so that others who come across this issue first check their env as it was the problem in my case. You have a more complicated case where you cannot switch to `node` from `jsdom`, in my case I should have switched to `node` in the first place as I was writing tests for the server (and for my client side code I have a separate jest config with env set to `jsdom`).

Unfortunately I don't know how to help you out, I would roll jest back to 26 and hope somebody does a fix in response to this issue. Yup, it is often difficult to migrate to a new major of some library, we have all been there

**Comment by davidmoshal** [comment]
Thanks

**Comment by roshkins** [comment]
When I put 
```typescript
/*
* @jest-environment node
*/
```
above my import statements it fixed it for me, since it was a dependency that was using setImmediate. It didn't work when I put it below the statements.

**Comment by dtostes-meli** [comment]
I had the same issue after updating to `jest v27`.
For those who could not solve with any of the previous responses (like me). After some research I've solved with:

`jest.config.js`
``` js
module.exports = {
  ...
  setupFilesAfterEnv: [
    '<rootDir>/jest.env.js',
  ],
  ...
}
```

`jest.env.js`
``` js
global.setImmediate = jest.useRealTimers;
```

**Comment by silvenon** [comment]
@davidmoshal what kind of code are you testing that needs both DOM and Node environment to exist? Maybe you can work around the problem somehow, so if you post the code someone might be able to help.

**Comment by bengeorge** [comment]
@silvenon is probably right here.

I chose to replace `setImmediate(foo)` usages with `setTimeout(foo, 0)`, and for our tests it was adequate.

**Comment by jaimeagudo** [comment]
> I had the same issue after updating to `jest v27`. For those who could not solve with any of the previous responses (like me). After some research I've solved with:
> 
> `jest.config.js`
> 
> ```js
> module.exports = {
>   ...
>   setupFilesAfterEnv: [
>     '<rootDir>/jest.env.js',
>   ],
>   ...
> }
> ```
> 
> `jest.env.js`
> 

I am afraid I don't see how this would have worked for anyone 

> ```js
> global.setImmediate = jest.useRealTimers;
> ```

TLDR: What worked for me is 

```
global.setImmediate = global.setImmediate || ((fn, ...args) => global.setTimeout(fn, 0, ...args));
```

For those interested https://nodejs.dev/learn/understanding-setimmediate

**Comment by fatso83** [comment]
@jaimeagudo You are just replacing a missing setImmediate with a call to setTimeout. Did you try just [importing the setImmediate timer](https://github.com/prisma/prisma/issues/8558#issuecomment-1040378575)?

**Comment by nabeelfarid** [comment]
> If you are actually using `setImmediate` in your tests, but your test environment is actually `jsdom`, the cleanest approach is to simply import the dependency. Just add this at the top:
> 
> ```js
> /**
>  * @jest-environment jsdom
>  */
> 
> import {setImmediate} from 'timers'
> ```

The import statement doest get rid of the error ' ReferenceError: setImmediate is not defined'
But when I try to test wether setImmediate has been processed , it doesn't seem to work. E.g. the following test fails

      import { setImmediate } from 'timers';

      jest.useFakeTimers('legacy');
      const m = jest.fn();
      setImmediate(m);
      jest.runAllImmediates();
      expect(m).toHaveBeenCalled();

how do you force setImmediate to process with Jest ?

**Comment by fatso83** [comment]
@nabeelfarid My comment was just actually referring to fixing the issue about setImmediate not being defined _in the test_, not in any _implementation_ code. Importing it will just bring it into the scope of the test and no implementation code will know about it and it will _not be added to the global scope_. Jest only stubs the bits in the global scope, so if you want Jest to do anything about `global.setImmediate`, you would need to bind it to global, AFAIK. As in `global.setImmediate = setImmediate`. 

Still, I do not think your example makes much sense anyhow: you only test the effects of what is happening inside your test module. This does not prove anything about code imported from outside your test module. `jsdom` does not have `setImmediate` in Jest 27, so any implementation code relying on it should fail regardless.

- https://github.com/facebook/jest/pull/11222
- https://github.com/testing-library/dom-testing-library/issues/914
- https://github.com/testing-library/react-testing-library/issues/918

**Comment by tobilg** [comment]
> `jest.env.js`
> 
> ```js
> global.setImmediate = jest.useRealTimers;
> ```

If you're using TypeScript, the following worked for me:

```typescript
global.setImmediate = jest.useRealTimers as unknown as typeof setImmediate;
```

**Comment by uroslates** [comment]
For me setting [testEnvironment](https://jestjs.io/docs/configuration#testenvironment-string) to `node` (`testEnvironment: 'node',`) within `jest.config.ts` solved the problem.

**Comment by regican** [comment]
Following @tobilg's solution, setting `global.setImmediate = global.setTimeout;` worked for me.

**Comment by kameshdhana** [comment]
> ```ts
> /*
> * @jest-environment node
> */
> ```

Such a banger workaround

**Comment by earonesty** [comment]
note:

this worked for me, and the bug is in `react-native-animated`.   it detects your environment as "node" even when you're running jsdom tests

`global.setImmediate = jest.useRealTimers;`

**Comment by jkomyno** [comment]
Some updates on this issue:
- we were not able to reproduce this problem with `jest@27.0.3` and `ts-jest@27.0.3`, using `prisma@5.1.0`
- we were able to run integration tests with Prisma Client on Jest with `testEnvironment: "node"`, using more recent versions (v29+)
- we ran those tests on Node.js v16.17.1 and on Node.js v18.16.1. We did not try using Node.js v14.17.4, as it is no longer compatible with Prisma.

What is the most recent version of Prisma and Node.js where you people have detected this issue? Moreover, I see several proposed solutions above that seem to have been appreciated by the community (in particular, setting the `testEnvironment: "node"` configuration property for Jest. Would you thus consider this issue as fixed?

**Comment by robbporto** [comment]
This is still happening.

**Comment by janpio** [comment]
@robbporto Can you provide some more information on the versions and exactly what you are doing so we can reproduce this, please?

**Comment by millsp** [comment]
@jkomyno have you tried it with the `jsdom` jest environment as well?

**Comment by maximgeerinck** [comment]
> global.setImmediate = jest.useRealTimers;



> I am afraid I don't see how this would have worked for anyone
> 
> > ```js
> > global.setImmediate = jest.useRealTimers;
> > ```
> 
> TLDR: What worked for me is
> 
> ```
> global.setImmediate = global.setImmediate || ((fn, ...args) => global.setTimeout(fn, 0, ...args));
> ```
> 
> For those interested https://nodejs.dev/learn/understanding-setimmediate

This is the only correct solution for me. the other one (`global.setImmediate = jest.useRealTimers;`) gave me timeouts.

**Comment by fatehMohamed14** [comment]
This is what worked for me (react typescript app)
`// @ts-ignore
window.setImmediate = window.setTimeout`

**Comment by juanhurtado4** [comment]
> This is what worked for me (react typescript app), add it inside the test file, socketio was causing the error for me
> 
> ```js
>     //@ts-ignore
>     window.setImmediate = window.setTimeout
> ```

I am also building a react typescript app that is using socketio. `window.setImmediate = window.setTimeout` worked for me. Another solution is to mock socketio `jest.mock('socket.io-client', () => SocketIoMock);`

**Comment by karim-pegasys** [comment]
> If you are actually using `setImmediate` in your tests, but your test environment is actually `jsdom`, the cleanest approach is to simply import the dependency. Just add this at the top:
> 
> ```js
> /**
>  * @jest-environment jsdom
>  */
> 
> import {setImmediate} from 'timers'
> ```

Using this then adding `global.setImmediate = setImmediate` right under it worked for me!

**Comment by dezsoab** [comment]
> > If you are actually using `setImmediate` in your tests, but your test environment is actually `jsdom`, the cleanest approach is to simply import the dependency. Just add this at the top:
> > ```js
> > /**
> >  * @jest-environment jsdom
> >  */
> > 
> > import {setImmediate} from 'timers'
> > ```
> 
> Using this then adding `global.setImmediate = setImmediate` right under it worked for me!

thank you! This solved for me

**Comment by YuriVini** [comment]
This worked for me: 

```
beforeEach(() => {
    jest.useFakeTimers();

    const mockAnimation = {
      start: jest.fn(),
      stop: jest.fn(),
      reset: jest.fn(),
    };

    jest.spyOn(Animated, "timing").mockImplementation(() => mockAnimation);
    jest.spyOn(Animated, "loop").mockImplementation((animation) => animation);
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.restoreAllMocks();
  });
```

**Comment by jkomyno** [comment]
Closing this, as a solution was already found: https://github.com/prisma/prisma/issues/8558#issuecomment-2192317371.
